unison-core/src/Unison/Term.hs

```diff
@@ -781,6 +782,23 @@ unLamsUntilDelay' t = case unLamsPred' (t, (/=) $ Var.named "()") of
   r@(Just _) -> r
   Nothing    -> Just ([], t)
 
+-- Same as unLamsUntilDelay', but only matches if the lambda body is a match
+-- expression, where the scrutinee is also the last argument of the lambda
+unLamsMatch'
+  :: Var v
+  => AnnotatedTerm2 vt at ap v a
+  -> Maybe ([v], [MatchCase ap (AnnotatedTerm2 vt at ap v a)])
+unLamsMatch' t = case unLamsUntilDelay' t of
+    Just (reverse -> (v1:vs), Match' (Var' v1') branches) |
+      (v1 == v1') && not (Set.member v1' (Set.unions $ freeVars <$> branches)) ->
+        Just (reverse vs, branches)
```

Comment:
These three lines are pretty epic and are doing a lot, but pretty easy to follow.

I might have written as `not (any (Set.member v1') (freeVars <$> branches))` to avoid needing to build up the union'd `Set` (which does a bunch of tree rebalancing stuff) but I think it's totally fine as is.